File: Graph/python/largestComponent.py
from .graph_utils import Graph


def largest_component(graph: Graph) -> int:
    visited: set[str] = set()
    largest = 0

    for node in graph:
        size = _explore_size(graph, node, visited)
        largest = max(largest, size)

    return largest


def _explore_size(graph: Graph, node: str, visited: set[str]) -> int:
    if node in visited:
        return 0
    visited.add(node)

    size = 1
    for neighbor in graph.get(node, []):
        size += _explore_size(graph, neighbor, visited)
    return size

File: Graph/ts/largestComponent.ts
import { Graph } from "./types";

export function largestComponent(graph: Graph): number {
  const visited = new Set<string>();
  let largest = 0;

  for (const node of Object.keys(graph)) {
    const size = exploreSize(graph, node, visited);
    largest = Math.max(largest, size);
  }

  return largest;
}

function exploreSize(graph: Graph, node: string, visited: Set<string>): number {
  if (visited.has(node)) return 0;
  visited.add(node);

  let size = 1;
  for (const neighbor of graph[node] ?? []) {
    size += exploreSize(graph, neighbor, visited);
  }
  return size;
}
